README.md
# WebKMS Context Repository _(webkms-context)_

[![Build status](https://img.shields.io/github/workflow/status/digitalbazaar/webkms-context/Node.js%20CI)](https://github.com/digitalbazaar/webkms-context/actions?query=workflow%3A%22Node.js+CI%22)
[![Coverage status](https://img.shields.io/codecov/c/github/digitalbazaar/webkms-context)](https://codecov.io/gh/digitalbazaar/webkms-context)
[![NPM Version](https://img.shields.io/npm/v/webkms-context.svg)](https://npm.im/webkms-context)

> A WebKMS context library for JavaScript.

## Table of Contents

- [Background](#background)
- [Install](#install)
- [Usage](#usage)
- [Commercial Support](#commercial-support)
- [License](#license)

## Background

See also (related specs):

* [WebKMS v0.1](https://w3c-ccg.github.io/webkms/)

## Install

Requires Node.js 12+

To install via NPM:

```
npm install webkms-context
```

## Usage

```js
const webkms = require('webkms-context');

webkms.CONTEXT_URL
// 'https://w3id.org/webkms/v1'

// Codec term map value for CBOR-LD
webkms.constants.CBORLD_CODEC_VALUE
// 0x19

// get context data for a specific context
webkms.CONTEXT
// full context object
```

This package can be used with bundlers, such as [webpack][], in browser
applications.

## API

The library exports the following properties:
- `CONTEXT_URL`
- `CONTEXT`
- `constants`: A Object that maps constants to well-known context URLs. The
  main constant `CONTEXT_URL` may be updated from time to time to the
  latest context location.
- `contexts`: A `Map` that maps URLs to full context data.
- `appContextMap`: For use with `cborld` library.

## Developing

WARNING: The `.jsonld` in `contexts/` is auto-generated by the `npm run build` script,
each time you run the test suite.

DO NOT edit it directly (or your changes will be quickly overwritten).
Instead, make all context changes to `js/context.js`.

## Commercial Support

Commercial support for this library is available upon request from
Digital Bazaar: support@digitalbazaar.com

## License

- BSD 3-Clause © Digital Bazaar
- See the [LICENSE](./LICENSE) file for details.

[webpack]: https://webpack.js.org/


CHANGELOG.md
# webkms-context ChangeLog

## 2.1.0 - 2022-01-14

### Added
- Add `maxCapabilityChainLength` term. This is a new term and not a
  breaking change. This term can be used to express the maximum
  acceptable length of a capability chain used in capability
  invocations at a particular invocation target (i.e., at a key URL).

## 2.0.0 - 2022-01-11

### Added
- **BREAKING**: Add `publicAlias` and `publicAliasTemplate` terms. The
  former is used to express a public alias (the public key identifier)
  for a key pair when generating a key. It may also be used to update
  the public key ID after the key has been generated. The other term,
  `publicAliasTemplate` serves a similar purpose, but instead holds
  a URI template value (per RFC 6570) that can be used during key
  generation to generate the public key ID from variables that appear
  in the key's public description.

### Changed
- **BREAKING**: Published package under `@digitalbazaar/webkms-context`.

## 1.0.0 - 2021-04-16

Initial version.


test/karma.conf.js
/*!
 * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';
module.exports = function(config) {
  const bundler = process.env.BUNDLER || 'webpack';
  const frameworks = ['mocha', 'chai'];

  const files = ['*.spec.js'];

  // browser launchers: https://npmjs.org/browse/keyword/karma-launcher
  // browsers: ['ChromeHeadless', 'Chrome', 'Firefox', 'Safari'],
  const browsers = ['ChromeHeadless'];

  const reporters = ['mocha'];
  const client = {
    mocha: {
      timeout: 10000, // 10 sec
      reporter: 'html'
      //delay: true
    }
  };

  // main bundle preprocessors
  const preprocessors = [];
  preprocessors.push(bundler);
  preprocessors.push('sourcemap');

  return config.set({
    frameworks,
    files,
    reporters,
    basePath: '',
    port: 9876,
    colors: true,
    browsers,
    client,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR ||
    //   config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    singleRun: true,

    // enable / disable watching file and executing test whenever any
    // file changes
    autoWatch: false,

    // preprocess matching files before serving them to the browser
    // available preprocessors:
    // https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      './*.js': preprocessors,
    },

    webpack: {
      devtool: 'inline-source-map',
      mode: 'development',
      module: {
        rules: [
          {
            test: /\.jsonld$/,
            loader: 'json-loader'
          }
        ]
      },
      node: {
        Buffer: false,
        crypto: false,
        util: false
      }
    }
  });
};


test/context.spec.js
/*!
 * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
 */
const chai = require('chai');
chai.should();
const {expect} = chai;

const {
  contexts, constants, appContextMap, CONTEXT_URL, CONTEXT
} = require('..');

describe('WebKMS Context', () => {
  it('constants', async () => {
    expect(appContextMap).to.exist;
    expect(constants).to.exist;
    expect(constants).to.have.property('CBORLD_VALUE');
    expect(CONTEXT_URL).to.exist;
    expect(CONTEXT).to.exist;
  });

  it('contexts', async () => {
    expect(contexts.get(CONTEXT_URL)).to.have.property('@context');
  });

  it('appContextMap', async () => {
    expect(appContextMap.get(CONTEXT_URL)).to.exist;
  });
});


js/context.js
/*!
 * Copyright (c) 2021-2022 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

module.exports = {
  '@context': {
    id: '@id',
    type: '@type',
    '@protected': true,

    DeleteKeyOperation: 'https://w3id.org/security#DeleteKeyOperation',
    DeriveSecretOperation: 'https://w3id.org/security#DeriveSecretOperation',
    ExportKeyOperation: 'https://w3id.org/security#ExportKeyOperation',
    GenerateKeyOperation: 'https://w3id.org/security#GenerateKeyOperation',
    KmsOperation: 'https://w3id.org/security#KmsOperation',
    RevokeKeyOperation: 'https://w3id.org/security#RevokeKeyOperation',
    SignOperation: 'https://w3id.org/security#SignOperation',
    UpdateKeyOperation: 'https://w3id.org/security#UpdateKeyOperation',
    UnwrapKeyOperation: 'https://w3id.org/security#UnwrapKeyOperation',
    VerifyOperation: 'https://w3id.org/security#VerifyOperation',
    WrapKeyOperation: 'https://w3id.org/security#WrapKeyOperation',
    invocationTarget: {
      '@id': 'https://w3id.org/security#invocationTarget', '@type': '@id'
    },
    kmsModule: 'https://w3id.org/security#kmsModule',
    maxCapabilityChainLength: {
      '@id': 'https://w3id.org/security#maxCapabilityChainLength',
      '@type': 'http://www.w3.org/2001/XMLSchema#integer'
    },
    publicKey: {
      '@id': 'https://w3id.org/security#publicKey', '@type': '@id'
    },
    publicAlias: {
      '@id': 'https://w3id.org/security#publicAlias',
      '@type': '@id'
    },
    publicAliasTemplate: 'https://w3id.org/security#publicAliasTemplate',
    revoked: {
      '@id': 'https://w3id.org/security#revoked',
      '@type': 'http://www.w3.org/2001/XMLSchema#dateTime'
    },
    signatureValue: 'https://w3id.org/security#signatureValue',
    unwrappedKey: 'https://w3id.org/security#unwrappedKey',
    verifyData: 'https://w3id.org/security#verifyData',
    wrappedKey: 'https://w3id.org/security#wrappedKey'
  }
};


karma.conf.js
/*!
 * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';
module.exports = function(config) {
  const bundler = process.env.BUNDLER || 'webpack';
  const frameworks = ['mocha', 'chai'];

  const files = ['*.spec.js'];

  // browser launchers: https://npmjs.org/browse/keyword/karma-launcher
  // browsers: ['ChromeHeadless', 'Chrome', 'Firefox', 'Safari'],
  const browsers = ['ChromeHeadless'];

  const reporters = ['mocha'];
  const client = {
    mocha: {
      timeout: 10000, // 10 sec
      reporter: 'html'
      //delay: true
    }
  };

  // main bundle preprocessors
  const preprocessors = [];
  preprocessors.push(bundler);
  preprocessors.push('sourcemap');

  return config.set({
    frameworks,
    files,
    reporters,
    basePath: 'test',
    port: 9876,
    colors: true,
    browsers,
    client,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR ||
    //   config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    singleRun: true,

    // enable / disable watching file and executing test whenever any
    // file changes
    autoWatch: false,

    // preprocess matching files before serving them to the browser
    // available preprocessors:
    // https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      './*.js': preprocessors,
    },

    webpack: {
      devtool: 'inline-source-map',
      mode: 'development',
      module: {
        rules: [
          {
            test: /\.jsonld$/,
            loader: 'json-loader'
          }
        ]
      },
      node: {
        Buffer: false,
        crypto: false,
        util: false
      }
    }
  });
};
